sentinel/src/lib/validation.js

```diff
@@ -1,290 +1,356 @@
-     * validation configs/separate rules in your settings. Example validation
-     * settings:
-     *
-     * [
-     *  {
-     *   property: "patient_id",
-     *   rule: "regex(\d{5})",
-     *   message: [{
-     *       content: "Patient ID must be 5 numbers.",
-     *       locale: "en"
-     *   }]
-     *  },
-     *  {
-     *   property: "patient_id",
-     *   rule: "unique('patient_id')",
-     *   message: [{
-     *       content: "Patient ID must be unique.",
-     *       locale: "en"
-     *   }]
-     *  }
-     * ]
-     *
-     * validate() modifies the property value of the second item to
-     * `patient_id_unique` so that pupil.validate() still returns a valid
-     * result.  Then we process the result once more to extract the custom
-     * validation results and error messages.
-     *
-     * @param ignores -
-     *   Optional array keys of doc that is always considered valid
-     *
-     * @callback Array of errors if validation failed, empty array otherwise.
-     * */
-    validate: function(doc, validations, ignores, callback) {
-        var self = module.exports,
-            result = {},
-            errors = [];
+    uniqueWithin: (doc, validation, callback) => {
+      const fields = _.clone(validation.funcArgs);
+      const duration = _parseDuration(fields.pop());
+      const startDate = moment()
+        .subtract(duration)
+        .valueOf();
+      _exists(doc, fields, { startDate: startDate }, (err, result) => {
+        if (err) {
+          logger.error('Error running "uniqueWithin" validation', err);
+        }
+        callback(err, !result);
+      });
+    },
+    exists: (doc, validation, callback) => {
+      const formName = validation.funcArgs[0];
+      const fieldName = validation.funcArgs[1];
+      _exists(
+        doc,
+        [fieldName],
+        { additionalFilter: `form:${formName}` },
+        (err, result) => {
+          if (err) {
+            logger.error('Error running "exists" validation', err);
+          }
+          callback(err, result);
+        }
+      );
+    },
+    /*
+         * Check if the week is a valid ISO week given a year.
+         */
+    isISOWeek: (doc, validation, callback) => {
+      const weekFieldName = validation.funcArgs[0];
+      const yearFieldName =
+        validation.funcArgs.length > 1 ? validation.funcArgs[1] : null;
+      if (
+        !_.has(doc, weekFieldName) ||
+        (yearFieldName && !_.has(doc, yearFieldName))
+      ) {
+        callback(
+          new Error('Validation failed: input field(s) do not exist'),
+          false
+        );
+      } else {
+        const year = yearFieldName
+          ? doc[yearFieldName]
+          : new Date().getFullYear();
+        const isValidISOWeek =
+          /^\d{1,2}$/.test(doc[weekFieldName]) &&
+          /^20\d{2}$/.test(year) &&
+          doc[weekFieldName] >= 1 &&
+          doc[weekFieldName] <= moment(`${year}-01-01`).isoWeeksInYear();
```

Comment:
Consider using `moment().year(year).isoWeeksInYear()` instead. It is easier to read and It will save you from needing to validate the year value - no errors are thrown if you pass an invalid value there, and it will default to current year. Proper validation of the year field should be done separately in configuration. 
If you consider validating the year value necessary, I don't think we should limit only to years starting with `20`, rather validate it's a 4 digit number.